### ISSUE TITLE
vault api does not support setting the increment by which a lease is renewed

### ISSUE THREAD

**Issue by jaloren** [hashicorp/vault#3679]
**Summary**
TLDR: when renewing a secret's lease, the api's Renew function does not allow you to specify the increment by which the secret's TTL is pushed into the future. Instead, after renewal, the lease TTL has been pushed into the future by the amount set in the vault system wide default ttl. If the following analysis is correct and this is a bug and my proposed solution is acceptable, I would be happy to submit a PR.

I have a go program that consumes the vault api. I am using the vault api to programmatically renew a secret's lease. Note that the lease duration is 5 minutes.

```golang
interval := 60
lease := "consul/creds/app/208303a0-7ba2-dfaa-b3db-d74f473054fa"
secret := &api.Secret{Renewable: true, LeaseID:   lease, LeaseDuration: 300}
renewerInput := &api.RenewerInput{Secret: secret,Grace: interval}
renewer, err := client.NewRenewer(renewerInput)
renewer.Renew()
```
However, after the Renew function runs, if I do a lookup on the lease, it shows that the lease will not expire for 768 hours (that happens  to be the default system TTL set for my vault installation). In addition, to not letting this be set, this lease TTL actually conflicts with the role TTL from which the secret was generated  -- that's 60 minutes. I would expect that the  newly renewed lease could be no more than the max TTL for the role that the secret is associated with i.e vault should have rejected the renewal as invalid. I am unsure if that's a bug or behaving as intended. 

I am reasonably confident that not being able to set the lease TTL on renewed lease  is a bug in the api. The file github.com/hashicorp/vault/api/renewer.go does has this line in the renewLease function.

```
renewal, err := client.Sys().Renew(leaseID, 0)
```
I tested this locally and verified that if i override the second argument with a non-zero value, then the renewed lease expires that amount of time (in seconds). 

My suggestion is to add a public int field to the RenewerInput called Increment and then in NewRenewer function, return the Renewer struct with the private field set like so:

```golang
&Renewer{
		client:  c,
		secret:  secret,
		grace:   grace,
                increment: i.increment
		random:  random,
		doneCh:  make(chan error, 1),
		renewCh: make(chan *RenewOutput, renewBuffer),
		stopped: false,
		stopCh:  make(chan struct{}),
	}
```
Then, in the renewLease function replace the 0 argument with the value from that struct.

```
renewal, err := client.Sys().Renew(leaseID, r.increment)
```

We would also want to document the effect of having the increment default to 0 (current behavior) on how the TTL increases when a secret is renewed.

**Environment:**
Centos 7.3

**Vault Version:**
Vault v0.9.0 ('bdac1854478538052ba5b7ec9a9ec688d35a3335')

**Expected Behavior:**
when the Renew function is invoked to renew a a secret lease, the duration of the renewed lease matches what was set in the LeaseDuration field in the api.Secret struct.

**Actual Behavior:**
when the Renew function is invoked, the  duration of the renewed leas is the default system ttl configured in vault.

**Comment by jefferai** [comment]
> when renewing a secret's lease, the api's Renew function does not allow you to specify the increment by which the secret's TTL is pushed into the future

This is correct. Currently, only token renewal specifies setting an increment.

> Instead, after renewal, the lease TTL has been pushed into the future by the amount set in the vault system wide default ttl.

This isn't quite accurate. It will fall back to the system-wide default TTL only if there have not been more narrowly-scoped default TTLs set, either on the mount (via mount-tune) or in the particular backend itself (most allow setting TTLs on a backend or role level).

**Comment by jaloren** [comment]
@jefferai  I am using the consul secret backend. I've created a role named 'zookeeper' on that backend with the lease of 1 hour.  

```
vault read consul/roles/zookeeper
Key       	Value
---       	-----
lease     	1h0m0s
policy    	c2VydmljZSAiem9va2VlcGVyIiB7IHBvbGljeSA9ICJ3cml0ZSJ9Cg==
token_type	client
```

I get a consul acl token by reading from the vault path  consul/creds/zookeeper and subsequently renew the lease. When I then perform a lease lookup, the TTL is 768 hours. As mentioned before, that behavior seemed wrong to me. Based on what you are saying above, this then is a bug, yes?

**Comment by jefferai** [comment]
It looks like the Consul backend is lacking support for this. Nearly every other backend should behave the way I described though :-/

**Comment by jefferai** [comment]
To be clear -- the Consul backend will currently honor the default lease tuned on a mount of the backend. It just won't use the role's set lease.
